Mage/src/main/java/mage/abilities/effects/common/continuous/GainAbilityControlledSpellsEffect.java

```diff
@@ -41,50 +41,52 @@ public GainAbilityControlledSpellsEffect copy() {
+            if (filter.match(card, game)) {
+                game.getState().addOtherAbility(card, ability);
             }
-            for (Card card : player.getHand().getCards(game)) {
-                if (filter.match(card, game)) {
-                    game.getState().addOtherAbility(card, ability);
-                }
+        }
+        for (Card card : player.getHand().getCards(game)) {
+            if (filter.match(card, game)) {
+                game.getState().addOtherAbility(card, ability);
             }
-            for (Card card : player.getGraveyard().getCards(game)) {
-                if (filter.match(card, game)) {
-                    game.getState().addOtherAbility(card, ability);
-                }
+        }
+        for (Card card : player.getGraveyard().getCards(game)) {
+            if (filter.match(card, game)) {
+                game.getState().addOtherAbility(card, ability);
             }
+        }
 
-            // workaround to gain cost reduction abilities to commanders before cast (make it playable)
-            game.getCommanderCardsFromCommandZone(player, CommanderCardType.ANY)
-                    .stream()
-                    .filter(card -> filter.match(card, game))
-                    .forEach(card -> {
-                        game.getState().addOtherAbility(card, ability);
-                    });
+        // workaround to gain cost reduction abilities to commanders before cast (make it playable)
+        game.getCommanderCardsFromCommandZone(player, CommanderCardType.ANY)
+                .stream()
+                .filter(card -> filter.match(card, game))
+                .forEach(card -> {
+                    game.getState().addOtherAbility(card, ability);
+                });
 
-            for (StackObject stackObject : game.getStack()) {
-                // only spells cast, so no copies of spells
-                if ((stackObject instanceof Spell)
-                        && !stackObject.isCopy()
-                        && stackObject.isControlledBy(source.getControllerId())) {
-                    Card card = game.getCard(stackObject.getSourceId());
-                    if (filter.match(card, game)) {
-                        game.getState().addOtherAbility(card, ability);
-                        return true;
-                    }
+        for (StackObject stackObject : game.getStack()) {
+            // only spells cast, so no copies of spells
+            if ((stackObject instanceof Spell)
+                    && !stackObject.isCopy()
+                    && stackObject.isControlledBy(source.getControllerId())) {
+                Card card = game.getCard(stackObject.getSourceId());
+                if (filter.match(card, game)) {
+                    game.getState().addOtherAbility(card, ability);
+                    return true;
                 }
             }
         }
-        return false;
+        return false; // TODO: Why is this returning false?
```

Comment:
Why is GainAbilityControlledSpellsEffect.apply returning false unless the ability is added to a single stack object?